### DISCUSSION TITLE
Hovers on keywords

### DISCUSSION THREAD

**Discussion by ydaveluy** [eclipse-langium/langium#1603] [Show and tell]
Hello,

I wanted to implement hovers on keywords like in [Xtext](https://blogs.itemis.com/en/xtext-usability-hovers-on-keywords).

I found a very simple way to do it:

- Write directly the documentation before keywords in langium grammar:
```langium
grammar HelloWorld


entry Model:
    (persons+=Person | greetings+=Greeting)*;

Person:
    /** define a new person*/
    'person' name=ID;

Greeting:
    /** say hello to a person*/
    'Hello' person=[Person:ID] '!';

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

- Override the MultilineCommentHoverProvider:
```typescript
export class MyDslHoverProvider extends MultilineCommentHoverProvider {

    override getHoverContent(document: LangiumDocument, params: HoverParams): MaybePromise<Hover | undefined> {
        const rootNode = document.parseResult?.value?.$cstNode;
        if (rootNode) {
            const offset = document.textDocument.offsetAt(params.position);

            const cstNode = CstUtils.findDeclarationNodeAtOffset(rootNode, offset, this.grammarConfig.nameRegexp);
            if (cstNode && cstNode.offset + cstNode.length > offset) {
                const targetNode = this.references.findDeclaration(cstNode);
                if (targetNode) {
                    return this.getAstNodeHoverContent(targetNode);
                }
                // add support for documentation on keywords
                if (GrammarAST.isKeyword(cstNode.grammarSource)) {
                    if (typeof (cstNode.grammarSource as AstNodeWithComment).$comment === 'string')
                        return this.getAstNodeHoverContent(cstNode.grammarSource)
                    else
                        // sometimes the comment is on the container
                        return this.getAstNodeHoverContent(cstNode.grammarSource.$container)
                }
            }
        }
        return undefined;
    }
}
```
- Inject MyDslHoverProvider
```typescript
export const MyDslModule: Module<MyDslServices, PartialLangiumServices & MyDslAddedServices> = {
    lsp: {
          HoverProvider: (services) => new MyDslHoverProvider(services),
    }
};

```

That's all !

Depending on how the grammar is written the command `this.getAstNodeHoverContent(cstNode.grammarSource)` does not always find the documentation. I would expect that a Keyword prefixed with a comment would be of  type AstNodeWithComment all the time but no. I don't know if it is expected or a bug.

This implementation assume that MyDsl grammar and langium grammars uses the same terminal rule for ML_COMMENT.

Maybe this functionality could be integrated by default in langium ?

**Comment by blewis13** [comment]
What are your imports for HoverParams and Hover? VSCode is offering up to 8 suggestions for each one.

**Reply by blewis13** [reply]
This seems to work for me. 

import { Hover, HoverParams } from "vscode-languageserver";

**Reply by msujew** [reply]
Using `vscode-languageserver` generally always works. These types in particular come from `vscode-languageserver-types` which is reexported from the other libraries.

**Reply by ydaveluy** [reply]
Yes I confirm, these includes are correct.
